        <div>
            <span class="notice">Read QC analysis is generated by subsampling fastq to [_SUBSAMPLE-RATE_]% of the reads.</span> <br />
            <span class="notice">Average Base Quality, Average Read Quality, Read GC and Cycle Nucleotide Composition are generated by the BBTools reformat.sh command.</span>
        </div>
        <h2  class="section-title">Average Base Quality</h2>
        <div class="section">
            <ul>
                <!--li>Overall base average quality score and the average quality score of base with quality score >= Q30.</li-->
                <li>The table contains the percentage of bases with a given quality score.</li>
            </ul>
            <table>
                <tr>
                    <td><span class="title-level-3">Overall Base Quality Score: </span><br /></td>
                    <td><span class="title-level-3">[_BASE-QUALITY-SCORE_] &plusmn; [_BASE-QUALITY-SCORE-STD_]</span></td>
                </tr>
                <tr>
                    <td><span class="title-level-3">Q30 Base Quality Score: </span></td>
                    <td><span class="title-level-3">[_Q30-BASE-QUALITY-SCORE_] &plusmn; [_Q30-BASE-QUALITY-SCORE-STD_]</span></td>
                </tr>
            </table>

            <table class='data'>
                <tr>
                    <th>Quality Value Threshold</th>
                    <th>&ge;Q30</th>
                    <th>&ge;Q25</th>
                    <th>&ge;Q20</th>
                    <th>&ge;Q15</th>
                    <th>&ge;Q10</th>
                    <th>&ge;Q5</th>
                </tr>
                <tr>
                    <th>Count of Bases</th>
                    <td>[_COUNT-OF-BAESE-Q30_]</td>
                    <td>[_COUNT-OF-BAESE-Q25_]</td>
                    <td>[_COUNT-OF-BAESE-Q20_]</td>
                    <td>[_COUNT-OF-BAESE-Q15_]</td>
                    <td>[_COUNT-OF-BAESE-Q10_]</td>
                    <td>[_COUNT-OF-BAESE-Q5_]</td>
                </tr>
                <tr>
                    <th>Percentage %</th>
                    <td>[_PCT-OF-BAESE-Q30_]</td>
                    <td>[_PCT-OF-BAESE-Q25_]</td>
                    <td>[_PCT-OF-BAESE-Q20_]</td>
                    <td>[_PCT-OF-BAESE-Q15_]</td>
                    <td>[_PCT-OF-BAESE-Q10_]</td>
                    <td>[_PCT-OF-BAESE-Q5_]</td>
                </tr>
            </table>
        </div>

        <h2 class="section-title">Average Read Quality</h2>
        <div class="section">
            <ul>
                <li>The data file contains the percentage of reads with a given average read quality.</li>
                <li>The plot is a histogram of average read quality.</li>
                <!--li>[_SUBSAMPLE-RATE_]% randomly sampled reads are used for this analysis.</li-->
            </ul>
            <table class='data'>
                <tr>
                    <th>Quality Value Threshold</th>
                    <th>&ge;Q30</th>
                    <th>&ge;Q25</th>
                    <th>&ge;Q20</th>
                    <th>&ge;Q15</th>
                    <th>&ge;Q10</th>
                    <th>&ge;Q5</th>
                </tr>
                <tr>
                    <th>Percentage of Reads %</th>
                    <td>[_PCT-OF-READS-Q30_]</td>
                    <td>[_PCT-OF-READS-Q25_]</td>
                    <td>[_PCT-OF-READS-Q20_]</td>
                    <td>[_PCT-OF-READS-Q15_]</td>
                    <td>[_PCT-OF-READS-Q10_]</td>
                    <td>[_PCT-OF-READS-Q5_]</td>
                </tr>
            </table>
            <br />
            <table class="plot">
                <tr><td style="text-align: center; font-family: sans-serif; font-size: 12px; border: none">Average Read Quality Histogram ([_AVG-READ-QUAL-HISTO-DATA_], [_AVG-READ-QUAL-HOSTO-D3_] )</td></tr>
                <tr><td><img src="[_AVG-READ-QUALITY-HISTOGRAM_]" alt="Average Read Quality Histogram Not Available"></td></tr>
            </table>
        </div>

        <h2 class="section-title">Average Base Position Quality</h2>
        <div class="section">
            <ul>
                <li>The data file contains the percentage of reads with a given average read quality.</li>
                <li>The plot shows average quality at each base position/cycle.</li>
                <!--li>[_SUBSAMPLE-RATE_]% randomly sampled reads are used for this analysis.</li-->
            </ul>
            <span class="title-level-3">Q20 for Read 1: [_READ_Q20_READ1_]</span> <br />
            <span class="title-level-3">Q20 for Read 2: [_READ_Q20_READ1_]</span>
            <table class="plot">
                <tr><td style="text-align: center; font-family: sans-serif; font-size: 12px; border: none">Average Base Position Quality Plot ( [_AVG-BASE-POS-QUAL-HISTO-DATA_], [_AVG-BASE-POS-QUAL-HISTO-D3_] )</td></tr>
                <tr><td><img src="[_AVG-BASE-POSITION-QUALITY_]" alt="Average Base Position Quality Not Available"></td></tr>
            </table>
        </div>

        <h2 class="section-title">Insert Size</h2>
        <div class="section">
            <ul>
                <li>BBTools bbmerge.sh is used to perform this analysis. bbmerge merges paired reads into single reads by overlap detection. With sufficient coverage, bbmerge can also merge nonoverlapping reads using gapped kmers.</li>
                <li>The "percentage reads joined" is calculated by (number of joined reads/total number of reads * 100).</li>
                <li>Raw reads are used for the insert size calculation (no trimming or filtering).</li>
            </ul>
            <table>
                <tr><td style="text-align: center; font-family: sans-serif; font-size: 12px; border: none">
                    Insert Size Statistics
                </td></tr>
                <tr><td style="border: none">
                    <table class="data">
                        <tr>
                            <th>Percentage Reads Joined</th>
                            <th>Average</th>
                            <th>Stddev</th>
                            <th>Mode</th>
                        </tr>
                        <tr>
                            <td>[_PCT-READS-JOINED_]</td>
                            <td>[_PCT-READS-JOINED-AVG_]</td>
                            <td>[_PCT-READS-JOINED-STDDEV_]</td>
                            <td>[_PCT-READS-JOINED-MODE_]</td>
                        </tr>
                    </table>
                </td></tr>
            </table>
            <br />
            <table class="plot">
                <tr><td style="text-align: center; font-family: sans-serif; font-size: 12px;">Insert size histogram ( [_INSERT-SIZE-HISTO-DATA_], [_INSERT-SIZE-HISTO-D3_] )</td></tr>
                <tr><td><img src="[_INSERT-SIZE-HISTOGRAM_]" alt="Insert Size Histogram Not Available"></td></tr>
            </table>
        </div>

<!--        <h3>End-of-read Illumina Adapter Check</h3>
        <div class="block">
            <span style="text-align: left; font-family: sans-serif; font-size: 12px;">The plot shows the location of Illumina adapters in reads</span>
            <br />
            <table>
                <tr><td style="text-align: center; font-family: sans-serif; font-size: 12px;">End-of-read Illumina Adapter Check (data file, interactive plot )</td></tr>
                <tr><td><img src="[_END-OF-READ-ILLUMINA-ADAPTER_]" alt="End-Of-Read Illumina Adapter Not Available"></td></tr>
            </table>
        </div>-->

        <h2 class="section-title">Read GC</h2>
        <div class="section">
            <ul>
                <li>Histogram of GC percents for subsampled reads.</li>
                <!--li>[_SUBSAMPLE-RATE_]% randomly sampled reads are used for this analysis.</li-->
            </ul>

            <span class="title-level-3">Average Read GC: [_READ-GC-AVG_] &plusmn; [_READ-GC-STDDEV_]%</span>
            <br />
            <br />
            <table class="plot">
                <tr><td style="text-align: center; font-family: sans-serif; font-size: 14px;border: none">Read GC Histogram ([_READ-QC-HISTO-DATA_], [_READ-QC-HISTO-D3_] )</td></tr>
                <tr><td><img src="[_READ-GC-HIST_]" alt="Read GC Not Available"></td></tr>
            </table>
        </div>

        <h2 class="section-title">Cycle Nucleotide Composition</h2>
        <div class="section">
            <ul>
                <li>The plot(s) show nucleotide counts at each cycle position.</li>
                <li>1 plot is created for each read.</li>
                <!--li>[_SUBSAMPLE-RATE_]% randomly sampled reads are used for this analysis.</li-->
            </ul>
            <br />
            <table class="plot">
                <tr>
                    <td style="text-align: center; font-family: sans-serif; font-size: 12px;">Read 1 Nucleotide Composition Frequency Plot ([_NUC-COMP-FREQ-R1-DATA_], [_NUC-COMP-FREQ-R1-D3_])</td>
                    <td style="text-align: center; font-family: sans-serif; font-size: 12px;">Read 2 Nucleotide Composition Frequency Plot ([_NUC-COMP-FREQ-R2-DATA_], [_NUC-COMP-FREQ-R2-D3_])</td>
                </tr>
                <tr>
                    <td><img src="[_CYCLE-NUCL-COMPOSITION-READ1_]" alt="Cycle Nucleotide Composition Read 1 Not Available"></td>
                    <td><img src="[_CYCLE-NUCL-COMPOSITION-READ2_]" alt="Cycle Nucleotide Composition Read 2 Not Available"></td>
                </tr>
            </table>
        </div>

        <h2 class="section-title">Percentage of Common Contaminants</h2>
        <div class="section">
            <ul>
                <li>The percentage of common contaminants in the randomly sampled reads are calculated using BBTools seal.sh and reported in the table.
                Phix: PhiX control sequence.</li>
                <!--li>Artifacts: Illumina sequencing artifacts and adapters not including DNA or RNA spike-in JGI Contaminants: Illumina sequencing adapters and artifacts and other various contaminants</li>
                <li>For "Artifacts (first 50bp)", The 1st 50bp (20bp for smRNA) of 2.68% randomly sampled reads are aligned to each other using seal.</li>
                <li>Non-synthetic Contaminants: prokaryotic and eukaryotic systematic contaminants</li>
                <li>Synthetic Contaminants: synthetic seqs used in library construction appearing as systematic contaminants</li>
                <li>Adapters: Illumina sequencing adapters Non-Synthetic Contaminants: common microbial contamination found in the sequencing process</li>
                <li>Additional info: <a href='https://docs.google.com/document/d/16A_bAIWjhQSM5_oIPlJJgNny8wU4hvKgVbBLGrI7x3w/edit#heading=h.ibfh0cow3oh2' targe='_blank'>Microbe Read Filtering: SOP 1077</a></li-->

            </ul>
            <!-- DO WE NEED TO CHECK illumina_read_percent_contamination_artifact not defined here? -->
            <table id="tag" class="data">
                <tr><th>Contaminant</th><th>Value</th><th>Contamination DB Description</th></tr>

                <tr><th>Artifacts</th><td class="right"><a href='[_CONTAM-ART-SEAL_]' target='_blank'>[_CONTAM-ART-SEA-PCT_]&#37; </a></td>
                    <td class="left">Illumina sequencing artifacts and adapters not including DNA or RNA spike-ins</td>
                </tr>

                [_CONTAN-ART-SEAL-FIRST-BP_]

                <tr><th>DNA spike-ins</th><td class="right"><a href='[_DNA-SPIKEIN-SEAL_]' target='_blank'>[_DNA-SPIKEIN-SEAL_PCT_]&#37; </a></td>
                    <td class="left">DNA spike-in sequences</td>
                </tr>
                <tr><th>RNA spike-ins</th><td class="right"><a href='[_RNA-SPIKEIN-SEAL_]' target='_blank'>[_RNA-SPIKEIN-SEAL_PCT_]&#37; </a></td>
                    <td class="left">RNA spike-in sequences</td>
                </tr>

                <tr><th>Fosmid</th><td class="right"><a href='[_CONTAM-FOSMID-SEAL_]' target='_blank'>[_CONTAM-FOSMID-SEAL-PCT_]&#37; </a></td>
                    <td class="left">fosmid sequences</td>
                </tr>
                <tr><th>Mitochondria</th><td class="right"><a href='[_CONTAM-MITO-SEAL_]' target='_blank'>[_CONTAM-MITO-SEAL-PCT_]&#37; </a></td>
                    <td class="left">mitochondria sequences</td>
                </tr>
                <tr><th>Chloroplast</th><td class="right"><a href='[_CONTAM-CHLO-SEAL_]' target='_blank'>[_CONTAM-CHLO-SEAL-PCT_]&#37; </a></td>
                    <td class="left">chloroplast sequences</td>
                </tr>
                <tr><th>PhiX</th><td class="right"><a href='[_CONTAM-PHIX-SEAL_]' target='_blank'>[_CONTAM-PHIX-SEAL-PCT_]&#37; </a></td>
                    <td class="left">control sequences</td>
                </tr>
                <tr><th>rRNA</th><td class="right"><a href='[_CONTAM-RRNA-SEAL_]' target='_blank'>[_CONTAM-RRNA-SEAL-PCT_]&#37; </a></td>
                    <td class="left">ribosomal RNA sequences</td>
                </tr>


                <tr><th>Non-synthetic Contaminants</th><td class="right"><a href='[_CONTAM-NON-SYN-SEAL_]' target='_blank'>[_CONTAM-NON-SYN-SEAL-PCT_]&#37; </a></td>
                    <td class="left">common microbial contamination found in the sequencing process</td>
                </tr>

                <tr><th class="left">Synthetic Contaminants</th><td class="right"><a href='[_CONTAM-SYN-SEAL_]' target='_blank'>[_CONTAM-SYN-SEAL-PCT_]&#37; </a></td>
                    <td class="left">synthetic sequencess used in library construction appearing as systematic contaminants</td>
                </tr>

                <!--tr><th>Adapters</th><td class="right"><a href='[_CONTAM-ADAPTER-SEAL_]' target='_blank'>[_CONTAM-ADAPTER-PCT_]&#37; </a></td>
                    <td>etc</td-->
                </tr>

            </table>
        </div>

        [_SKETCH-TABLE_]